src/engine/controls/bpmcontrol.cpp

```diff
@@ -117,6 +143,34 @@ BpmControl::BpmControl(const QString& group,
             this,
             &BpmControl::slotTranslateBeatsMove,
             Qt::DirectConnection);
+    m_pBeatsSetMarker = std::make_unique<ControlPushButton>(
+            ConfigKey(group, "beats_set_marker"), false);
```

Comment:
This is not the same. You don't need to set every bar line manually, they are interpolated based on the position of the marker and the BPM. The "Marker" is the frame position where the first downbeat with *new* settings is. I think the term is fine.

Reply:
> You don't need to set every bar line manually 

That's clear, but you also set at least the first. And it is about the alignment of the bars.
So I think my proposal is not completely wrong. 

Is the an other way to relate "beats_set_marker" to bars? 
Maybe "beats_start_new_bars" or something.

Reply:
I think it makes sense to keep the clear distinction between markers and bars/downbeat as @Holzhaus rightfully pointed out.
There is is also the fact that marker define a new grid, so in the case where the marker is set in between two beats, the new grid will start of tempo compare to the previous grid. As a reminder, this is a requirement to deal with tracks that have variable tempo or may shift by a bit fraction in the track.

Reply:
Ah OK, in is just a new grid start bar line. I thought this sets a bar line and then other bar lines are interpolated form that. 

The essence of my first comment is that "beats_set_marker" reads as if you set a beat marker which is just the vertical line we had always marking beats. I think this is a real issue and needs to be solved. 

proposals: 
"beats_set_bar_change"
"beat_signature_change" 
"beats_set_bars_start"

Reply:
Happy with `beats_mark_bar_change` - does that work for you @Holzhaus ?

Reply:
Maybe I've not understood everything and I'm not a native English speaker, but `beats_mark_bar_change` sounds strange to me. I think `time_signature_change_marker` or `time_signature_changing_beat` would be more self explaining.

Reply:
I agree with @JoergAtGithub that this sounds weird. `time_signature_change_marker` is also not a good fit, because this is *not about time signatures*. You can set a marker and have the same time signature as before (e.g., 4/4) - to be honest, i think that would be case in the vast majority of cases, because changing time signatures are relatively rare compared to only changing BPM in my experience.

FYI in Serato Pro, this are also called "marker", as you can see in the editing controls:
![](https://support.serato.com/hc/article_attachments/212301647/Screen_Shot_2016-09-20_at_3.43.16_pm.png)
In the manual, they call it "Beat Warp Marker" (although this term is not the GUI), see: https://support.serato.com/hc/en-us/articles/227626968-Beat-Warp-Markers

In my opionion, the term "marker" is fine. I would strongly prefer `beats_set_marker`. If you really want to be verbose, maybe `beats_set_change_marker` would also be okay, but all other suggestions so far seem less fitting to me.

Reply:
Sounds like we are trying to wrap up a non-trivial concept into a small cryptic string. Adding to the equation our different native languages, that just feels like an impossible task.

I think what this needs is a good documentation, then whatever key we agree with is kinda irrelevant. Personally, I never "guess" keys, I have the Mixxx Controls documents bookmarked and just search for what I need there.

So TL;DR, I'm happy with anything, but I'd vote for keeping what we currently have.

Reply:
Ok, I understand `beats_set_change_marker` is than a good fit. 

@Holzhaus, you post reveals another conceptual issue, it is worth noting: 

We have in one hand beat maps, where basically everything beat is a "change marker". On the other hand we have real "bar changes" (Taktwechsel) that would be annotated as a double bar line. Sometimes with a written measure signature. The same is used for a key change, along with new B's and Sharps. 

I would love to have that visualized somehow, but not clutter my old auto detected maps with this info. Do we have plans about distinguish these cases visually?

Reply:
Please stay focused on the topic! Key changes and general music annotation are completely unrelated to this PR.

Reply:
> We have in one hand beat maps, where basically everything beat is a \"change marker\". On the other hand we have real \"bar changes\" (Taktwechsel) that would be annotated as a double bar line.

Just to clarify: you are referring to the protobuf representation. Here we are talking about the user-facing/in-memory representation.

What my PR and this one do, is converting the protobuf representation into the in-memory "Beats" representation, which consists of one or more markers where each marker is a frame position + BPM.

For (const) beatgrids, there is only one change marker, and the beats before and behind the marker are interpolate. For (non-const) beatmaps, there is one or more of such markers. Insignificant drift between beats in the beatmap are ignored, so markers are only inferred for significant changes.

So no, not every beat from a beatmap is a "marker".

Reply:
That sounds good. 

I was talking about the visual presentation. 
What should happen visually if a change marker is set and what should happen if a beat map with change marker is loaded. 
My proposal is to mark them conditionally with a double bar line. The conditions are however not clear. 

So the question is, do we need a control to set this visual double bar line, or will there be a magic formula to do it?

Reply:
This PR doesn't contain any visuals, so di'm not sure what you're suggesting.

Let's please not extend the scope of this PR to visuals when @acombier already worked hard to split the visual stuff from this PR to achieve a reviewable diff size.

Reply:
The visuals are implications here. 

Right, here we are looking to the CO interface. 
I'm interested in the future plans to align our requirements. 

As I have understood with the new "beats_set_change_marker" CO we can adjust either adjust tempo slightly, or mark a signature change which is a double bar line on a musical sheet. 

I consider this to be sufficient for all use cases, in a way that all change markers shall be visible on the waveform. But I like to confirm it with you to avoid future discussions. (Compared to an additional interface which allows to move single beats) 

We have two use cases regarding tempo changes: 
1. A const beat grid which aligns at the start and end but is off in the middle. 
 * User can find the tempo change positions, preferably pick a downbeat not on the grid and mark it using the new control. The beast before and after, schall be adjusted to hit this new beat marked with e.g a double bar line. 
2. A non const beat grid where the beats are more likely on time but with issues making "tshank" instead of "boom". The latest detector is able to group beats with the same tempo. I consider it very much desired to not change tempo in every beat to avoid youling in followers. 
 * User may want to explicitly mark tempo changes and combine regions if similar tempo to one region. After finishing, it is probably desired to be in the same state as 1. User should also be able to align change markers to downbeats. Technical he is free to mark every beat, but that should be an exception. 

Can you confirm that? 

I think Mixxx should make it hard to create a padding bar, that does not match the previous and not the follower bars. In case users really wants it, they can mark the start of this padding bar as well. 

What do you think?

Reply:
I guess we could make the marker align on the previously set grid if quantize is on. IMO that makes sense indeed to guide users to keep, their beatgrid as aligned as possible. 
Could we please address that as a follow up PR tho? I really would like to get this merge ASAP to set the base for future work related to beatgrid.

Reply:
I don't have demands for additional changes here, so that's fine. I just like to confirm that we are on the right track. @Holzhaus, can you confirm that as well?

Reply:
I have pushed the CO update. Are we happy to move forward now?

Reply:
Yes, thank you.